File: src/main/java/com/procurement/notice/service/RelatedProcessServiceImpl.java
package com.procurement.notice.service;

import com.datastax.driver.core.utils.UUIDs;
import com.procurement.notice.model.budget.EI;
import com.procurement.notice.model.budget.FS;
import com.procurement.notice.model.ocds.RelatedProcess;
import com.procurement.notice.model.ocds.RelatedProcessScheme;
import com.procurement.notice.model.ocds.RelatedProcessType;
import com.procurement.notice.model.tender.dto.CheckFsDto;
import com.procurement.notice.model.tender.ms.Ms;
import com.procurement.notice.model.tender.record.Record;
import java.util.Collections;
import java.util.LinkedHashSet;
import java.util.Objects;
import org.springframework.beans.factory.annotation.Value;
import org.springframework.stereotype.Service;

@Service
public class RelatedProcessServiceImpl implements RelatedProcessService {

    private static final String FS_SEPARATOR = "-FS-";
    private static final String URI_SEPARATOR = "/";

    @Value("${uri.budget}")
    private String budgetUri;
    @Value("${uri.tender}")
    private String tenderUri;

    @Override
    public void addFsRelatedProcessToEi(final EI ei, final String fsOcId) {
        final RelatedProcess relatedProcess = new RelatedProcess(
                UUIDs.timeBased().toString(),
                Collections.singletonList(RelatedProcessType.X_FINANCE_SOURCE),
                RelatedProcessScheme.OCID,
                fsOcId,
                getBudgetUri(ei.getOcid(), fsOcId)
        );
        if (Objects.isNull(ei.getRelatedProcesses())) {
            ei.setRelatedProcesses(new LinkedHashSet<>());
        }
        ei.getRelatedProcesses().add(relatedProcess);
    }

    @Override
    public void addEiRelatedProcessToFs(final FS fs, final String eiOcId) {
        final RelatedProcess relatedProcess = new RelatedProcess(
                UUIDs.timeBased().toString(),
                Collections.singletonList(RelatedProcessType.PARENT),
                RelatedProcessScheme.OCID,
                eiOcId,
                getBudgetUri(eiOcId, eiOcId)
        );
        if (Objects.isNull(fs.getRelatedProcesses())) {
            fs.setRelatedProcesses(new LinkedHashSet<>());
        }
        fs.getRelatedProcesses().add(relatedProcess);
    }

    @Override
    public void addMsRelatedProcessToEi(final EI ei, final String msOcId) {
        final RelatedProcess relatedProcess = new RelatedProcess(
                UUIDs.timeBased().toString(),
                Collections.singletonList(RelatedProcessType.X_EXECUTION),
                RelatedProcessScheme.OCID,
                msOcId,
                getTenderUri(msOcId, msOcId)
        );
        if (Objects.isNull(ei.getRelatedProcesses())) {
            ei.setRelatedProcesses(new LinkedHashSet<>());
        }
        ei.getRelatedProcesses().add(relatedProcess);
    }

    @Override
    public void addMsRelatedProcessToFs(final FS fs, final String msOcId) {
        final RelatedProcess relatedProcess = new RelatedProcess(
                UUIDs.timeBased().toString(),
                Collections.singletonList(RelatedProcessType.X_EXECUTION),
                RelatedProcessScheme.OCID,
                msOcId,
                getTenderUri(msOcId, msOcId)
        );
        if (Objects.isNull(fs.getRelatedProcesses())) {
            fs.setRelatedProcesses(new LinkedHashSet<>());
        }
        fs.getRelatedProcesses().add(relatedProcess);
    }

    @Override
    public void addEiFsRecordRelatedProcessToMs(final Ms ms,
                                                final CheckFsDto checkFs,
                                                final String ocId,
                                                final RelatedProcessType recordProcessType) {
        if (Objects.isNull(ms.getRelatedProcesses())) {
            ms.setRelatedProcesses(new LinkedHashSet<>());
        }
        /*record*/
        ms.getRelatedProcesses().add(new RelatedProcess(
                UUIDs.timeBased().toString(),
                Collections.singletonList(recordProcessType),
                RelatedProcessScheme.OCID,
                ocId,
                getTenderUri(ms.getOcid(), ocId)));
        /*expenditure items*/
        checkFs.getEi().forEach(eiCpId ->
                ms.getRelatedProcesses().add(new RelatedProcess(
                        UUIDs.timeBased().toString(),
                        Collections.singletonList(RelatedProcessType.X_EXPENDITURE_ITEM),
                        RelatedProcessScheme.OCID,
                        eiCpId,
                        getBudgetUri(eiCpId, eiCpId))
                )
        );
        /*financial sources*/
        ms.getPlanning().getBudget().getBudgetBreakdown().forEach(fs ->
                ms.getRelatedProcesses().add(new RelatedProcess(
                        UUIDs.timeBased().toString(),
                        Collections.singletonList(RelatedProcessType.X_BUDGET),
                        RelatedProcessScheme.OCID,
                        fs.getId(),
                        getBudgetUri(getEiCpIdFromOcId(fs.getId()), fs.getId()))
                )
        );
    }

    @Override
    public void addMsRelatedProcessToRecord(final Record record, final String msOcId) {
        /*ms*/
        final RelatedProcess relatedProcess = new RelatedProcess(
                UUIDs.timeBased().toString(),
                Collections.singletonList(RelatedProcessType.PARENT),
                RelatedProcessScheme.OCID,
                msOcId,
                getTenderUri(msOcId, msOcId));
        if (Objects.isNull(record.getRelatedProcesses())) {
            record.setRelatedProcesses(new LinkedHashSet<>());
        }
        record.getRelatedProcesses().add(relatedProcess);
    }

    @Override
    public void addRecordRelatedProcessToMs(final Record record, final String msOcId, final RelatedProcessType recordProcessType) {
        record.getRelatedProcesses()
                .add(new RelatedProcess(
                        UUIDs.timeBased().toString(),
                        Collections.singletonList(recordProcessType),
                        RelatedProcessScheme.OCID,
                        record.getOcid(),
                        getTenderUri(msOcId, record.getOcid())));
    }

    @Override
    public void addPervRecordRelatedProcessToRecord(final Record record, final String prevRecordOcId, final String msOcId) {
        record.getRelatedProcesses()
                .add(new RelatedProcess(
                        UUIDs.timeBased().toString(),
                        Collections.singletonList(RelatedProcessType.X_PRESELECTION),
                        RelatedProcessScheme.OCID,
                        prevRecordOcId,
                        getTenderUri(msOcId, prevRecordOcId)));
    }

    private String getEiCpIdFromOcId(final String ocId) {
        final int pos = ocId.indexOf(FS_SEPARATOR);
        return ocId.substring(0, pos);
    }

    private String getBudgetUri(final String cpId, final String ocId) {
        return budgetUri + cpId + URI_SEPARATOR + ocId;
    }

    private String getTenderUri(final String cpId, final String ocId) {
        return tenderUri + cpId + URI_SEPARATOR + ocId;
    }
}

File: src/main/kotlin/com/procurement/notice/service/RelatedProcessServiceImpl.kt
package com.procurement.notice.service

import com.datastax.driver.core.utils.UUIDs
import com.procurement.notice.model.budget.EI
import com.procurement.notice.model.budget.FS
import com.procurement.notice.model.ocds.RelatedProcess
import com.procurement.notice.model.ocds.RelatedProcessScheme
import com.procurement.notice.model.ocds.RelatedProcessType
import com.procurement.notice.model.tender.dto.CheckFsDto
import com.procurement.notice.model.tender.ms.Ms
import com.procurement.notice.model.tender.record.ContractRecord
import com.procurement.notice.model.tender.record.Record
import org.springframework.beans.factory.annotation.Value
import org.springframework.stereotype.Service

@Service
interface RelatedProcessService {

    fun addFsRelatedProcessToEi(ei: EI, fsOcId: String)

    fun addEiRelatedProcessToFs(fs: FS, eiOcId: String)

    fun addMsRelatedProcessToEi(ei: EI, msOcId: String)

    fun addMsRelatedProcessToFs(fs: FS, msOcId: String)

    fun addEiFsRecordRelatedProcessToMs(ms: Ms, checkFs: CheckFsDto, ocId: String, processType: RelatedProcessType)

    fun addMsRelatedProcessToRecord(record: Record, msOcId: String)

    fun addRecordRelatedProcessToMs(ms: Ms, recordOcId: String, processType: RelatedProcessType)

    fun addRecordRelatedProcessToRecord(record: Record, prevRecordOcId: String, msOcId: String, processType: RelatedProcessType)

    fun addMsRelatedProcessToContract(record: ContractRecord, msOcId: String)

    fun addRecordRelatedProcessToContract(record: ContractRecord, recordOcId: String, msOcId: String, processType: RelatedProcessType)

}

@Service
class RelatedProcessServiceImpl : RelatedProcessService {

    @Value("\${uri.budget}")
    private val budgetUri: String? = null
    @Value("\${uri.tender}")
    private val tenderUri: String? = null

    override fun addFsRelatedProcessToEi(ei: EI, fsOcId: String) {
        if (ei.relatedProcesses == null) ei.relatedProcesses = hashSetOf()
        ei.relatedProcesses?.add(RelatedProcess(
                id = UUIDs.timeBased().toString(),
                relationship = listOf(RelatedProcessType.X_FINANCE_SOURCE),
                scheme = RelatedProcessScheme.OCID,
                identifier = fsOcId,
                uri = getBudgetUri(ei.ocid, fsOcId)
        ))
    }

    override fun addEiRelatedProcessToFs(fs: FS, eiOcId: String) {
        if (fs.relatedProcesses == null) fs.relatedProcesses = hashSetOf()
        fs.relatedProcesses?.add(RelatedProcess(
                id = UUIDs.timeBased().toString(),
                relationship = listOf(RelatedProcessType.PARENT),
                scheme = RelatedProcessScheme.OCID,
                identifier = eiOcId,
                uri = getBudgetUri(eiOcId, eiOcId)
        ))
    }

    override fun addMsRelatedProcessToEi(ei: EI, msOcId: String) {
        if (ei.relatedProcesses == null) ei.relatedProcesses = hashSetOf()
        ei.relatedProcesses?.add(RelatedProcess(
                id = UUIDs.timeBased().toString(),
                relationship = listOf(RelatedProcessType.X_EXECUTION),
                scheme = RelatedProcessScheme.OCID,
                identifier = msOcId,
                uri = getTenderUri(msOcId, msOcId)
        ))
    }

    override fun addMsRelatedProcessToFs(fs: FS, msOcId: String) {
        if (fs.relatedProcesses == null) fs.relatedProcesses = hashSetOf()
        fs.relatedProcesses?.add(RelatedProcess(
                id = UUIDs.timeBased().toString(),
                relationship = listOf(RelatedProcessType.X_EXECUTION),
                scheme = RelatedProcessScheme.OCID,
                identifier = msOcId,
                uri = getTenderUri(msOcId, msOcId)
        ))
    }

    override fun addEiFsRecordRelatedProcessToMs(ms: Ms, checkFs: CheckFsDto, ocId: String, processType: RelatedProcessType) {
        if (ms.relatedProcesses == null) ms.relatedProcesses = hashSetOf()
        /*record*/
        ms.relatedProcesses?.add(RelatedProcess(
                id = UUIDs.timeBased().toString(),
                relationship = listOf(processType),
                scheme = RelatedProcessScheme.OCID,
                identifier = ocId,
                uri = getTenderUri(ms.ocid!!, ocId)))
        /*expenditure items*/
        checkFs.ei.asSequence().forEach { eiCpId ->
            ms.relatedProcesses?.add(RelatedProcess(
                    id = UUIDs.timeBased().toString(),
                    relationship = listOf(RelatedProcessType.X_EXPENDITURE_ITEM),
                    scheme = RelatedProcessScheme.OCID,
                    identifier = eiCpId,
                    uri = getBudgetUri(eiCpId, eiCpId))
            )
        }
        /*financial sources*/
        ms.planning?.budget?.budgetBreakdown?.asSequence()?.forEach {
            ms.relatedProcesses?.add(RelatedProcess(
                    id = UUIDs.timeBased().toString(),
                    relationship = listOf(RelatedProcessType.X_BUDGET),
                    scheme = RelatedProcessScheme.OCID,
                    identifier = it.id,
                    uri = getBudgetUri(getEiCpIdFromOcId(it.id!!), it.id)))
        }
    }

    override fun addMsRelatedProcessToRecord(record: Record, msOcId: String) {
        if (record.relatedProcesses == null) record.relatedProcesses = hashSetOf()
        record.relatedProcesses?.add(RelatedProcess(
                id = UUIDs.timeBased().toString(),
                relationship = listOf(RelatedProcessType.PARENT),
                scheme = RelatedProcessScheme.OCID,
                identifier = msOcId,
                uri = getTenderUri(msOcId, msOcId)))
    }

    override fun addRecordRelatedProcessToMs(ms: Ms, recordOcId: String, processType: RelatedProcessType) {
        if (ms.relatedProcesses == null) ms.relatedProcesses = hashSetOf()
        ms.relatedProcesses?.add(RelatedProcess(
                id = UUIDs.timeBased().toString(),
                relationship = listOf(processType),
                scheme = RelatedProcessScheme.OCID,
                identifier = recordOcId,
                uri = getTenderUri(ms.ocid!!, recordOcId)))
    }

    override fun addRecordRelatedProcessToRecord(record: Record, prevRecordOcId: String, msOcId: String, processType: RelatedProcessType) {
        if (record.relatedProcesses == null) record.relatedProcesses = hashSetOf()
        record.relatedProcesses?.add(RelatedProcess(
                id = UUIDs.timeBased().toString(),
                relationship = listOf(processType),
                scheme = RelatedProcessScheme.OCID,
                identifier = prevRecordOcId,
                uri = getTenderUri(msOcId, prevRecordOcId)))
    }

    override fun addMsRelatedProcessToContract(record: ContractRecord, msOcId: String) {
        if (record.relatedProcesses == null) record.relatedProcesses = hashSetOf()
        record.relatedProcesses?.add(RelatedProcess(
                id = UUIDs.timeBased().toString(),
                relationship = listOf(RelatedProcessType.PARENT),
                scheme = RelatedProcessScheme.OCID,
                identifier = msOcId,
                uri = getTenderUri(msOcId, msOcId)))
    }

    override fun addRecordRelatedProcessToContract(record: ContractRecord, recordOcId: String, msOcId: String,
                                                   processType: RelatedProcessType) {
        if (record.relatedProcesses == null) record.relatedProcesses = hashSetOf()
        record.relatedProcesses?.add(RelatedProcess(
                id = UUIDs.timeBased().toString(),
                relationship = listOf(processType),
                scheme = RelatedProcessScheme.OCID,
                identifier = recordOcId,
                uri = getTenderUri(msOcId, recordOcId)))
    }

    private fun getEiCpIdFromOcId(ocId: String): String {
        val pos = ocId.indexOf(FS_SEPARATOR)
        return ocId.substring(0, pos)
    }

    private fun getBudgetUri(cpId: String?, ocId: String?): String {
        return budgetUri + cpId + URI_SEPARATOR + ocId
    }

    private fun getTenderUri(cpId: String, ocId: String?): String {
        return tenderUri + cpId + URI_SEPARATOR + ocId
    }

    companion object {
        private val FS_SEPARATOR = "-FS-"
        private val URI_SEPARATOR = "/"
    }
}
